src/test/java/com/artipie/asto/StorageExclusivelyTest.java

```diff
@@ -0,0 +1,116 @@
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NON-INFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+package com.artipie.asto;
+
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionException;
+import java.util.concurrent.CompletionStage;
+import java.util.function.Function;
+import org.hamcrest.MatcherAssert;
+import org.hamcrest.core.IsInstanceOf;
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.TestTemplate;
+import org.junit.jupiter.api.extension.ExtendWith;
+
+/**
+ * Tests for {@link Storage#exclusively(Key, Function)}.
+ *
+ * @since 0.27
+ */
+@ExtendWith(StorageExtension.class)
+public final class StorageExclusivelyTest {
+
+    @TestTemplate
+    void shouldFailExclusivelyForSameKey(final Storage storage) {
+        final Key key = new Key.From("shouldFailConcurrentExclusivelyForSameKey");
+        final FakeOperation operation = new FakeOperation();
+        storage.exclusively(key, operation);
+        operation.started.join();
+        final CompletionException completion = Assertions.assertThrows(
+            CompletionException.class,
+            () -> storage.exclusively(key, new FakeOperation()).toCompletableFuture().join()
+        );
+        MatcherAssert.assertThat(
+            completion.getCause(),
+            new IsInstanceOf(IllegalStateException.class)
+        );
+    }
+
+    @TestTemplate
+    void shouldRunExclusivelyForDiffKey(final Storage storage) {
+        final Key one = new Key.From("shouldRunExclusivelyForDiffKey-1");
+        final Key two = new Key.From("shouldRunExclusivelyForDiffKey-2");
+        final FakeOperation operation = new FakeOperation();
+        storage.exclusively(one, operation);
+        operation.started.join();
+        Assertions.assertDoesNotThrow(
+            () -> storage.exclusively(two, new FakeOperation(CompletableFuture.allOf()))
+                .toCompletableFuture().join()
+        );
+    }
+
+    @TestTemplate
+    void shouldRunExclusivelyWhenPrevFinished(final Storage storage) {
+        final Key key = new Key.From("shouldRunExclusivelyWhenPrevFinished");
+        final FakeOperation operation = new FakeOperation(CompletableFuture.allOf());
+        storage.exclusively(key, operation).toCompletableFuture().join();
+        Assertions.assertDoesNotThrow(
+            () -> storage.exclusively(key, new FakeOperation(CompletableFuture.allOf()))
+                .toCompletableFuture().join()
+        );
+    }
+
+    /**
+     * Fake operation with controllable start and finish.
```

Comment:
@olegmoz I'd describe a bit more about this class, since it's not clear how to use it and what's `started` and `finished` fields here

Reply:
@g4s8 comments updated, hope they are clear now